### PR TITLE
[pigeon] Support other hosts in generated file CI checks

### PR DESCRIPTION
**Pull request by stuartmorgan**
Reworks Pigeon's CI validation of generated files to support multiple hosts, rather than only Linux, with the ability to set specific languages per host, to allow us to run validation for each language on whatever host it is most convenient to set up a formatter for that language, rather than having to support all languages on Linux.

See discussion in https://github.com/flutter/packages/pull/5928

Part of https://github.com/flutter/flutter/issues/41129

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
test-exempt: is a test

File: packages/pigeon/tool/run_tests.dart [hunk 1]
```diff
@@ -35,62 +36,56 @@ void _validateTestCoverage(List<List<String>> shards) {
...
+    languagesToValidate = <GeneratorLanguage>{
+      GeneratorLanguage.cpp,
+      GeneratorLanguage.dart,
+      GeneratorLanguage.java,
+      GeneratorLanguage.kotlin,
+      GeneratorLanguage.objc,
+    };
+  } else if (Platform.isMacOS) {
+    languagesToValidate = <GeneratorLanguage>{
+      // TODO(stuartmorgan): Enable this in or after https://github.com/flutter/packages/pull/5928
```
Comment (jmagman) [inline packages/pigeon/tool/run_tests.dart:76]: Merged! https://github.com/flutter/packages/pull/5928#discussion_r1459560316

**Review by jmagman — APPROVED** [review]

File: packages/pigeon/tool/run_tests.dart [hunk 2]
```diff
@@ -138,6 +136,17 @@ Future<void> _validateGeneratedExampleFiles() async {
...
 }
 
+Set<String> _extensionsForLanguage(GeneratorLanguage language) {
+  return switch (language) {
+    GeneratorLanguage.cpp => <String>{'cc', 'cpp', 'h'},
+    GeneratorLanguage.dart => <String>{'dart'},
+    GeneratorLanguage.java => <String>{'java'},
+    GeneratorLanguage.kotlin => <String>{'kt'},
+    GeneratorLanguage.swift => <String>{'swift'},
+    GeneratorLanguage.objc => <String>{'m', 'mm'},
```
Comment (tarrinneal) [inline packages/pigeon/tool/run_tests.dart:146]: how are we handling .h files for objc?

**Review by tarrinneal — APPROVED** [review]

File: packages/pigeon/tool/run_tests.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/pigeon/tool/run_tests.dart:146]: Oops! They should be listed here; having duplicates will be fine since.

**Comment by stuartmorgan** [comment]
changelog override: the only non-test change is formatting of an example file, which isn't going to be interesting to clients.

**Comment by stuartmorgan** [comment]
version override: quick_actions code changes are just formatting, so do not affect clients.

**Pull request merged**